File: BinarySearch/34.Search for a Range/Solution.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] res = new int[]{-1,-1};
        if(nums.length==0 || nums[0]>target || nums[nums.length-1]<target) {
            return res;
        }
        
        res[0] = getFirstOccurance(nums, target);
        res[1] = getLastOccurance(nums, target);
        
        return res;
        
    }
    
    public int getFirstOccurance(int[] nums,int target) {
        int left = 0;
        int right = nums.length - 1;
        while(left < right -1) {
            int mid = left + (right-left)/2;
           if( nums[mid]<target) {
                left = mid;
            } else {
                right = mid;
            }
        }
        if(nums[left] == target) {
            return left;
        }
        if(nums[right] == target) {
            return right;
        }
        return -1;
    }
    
     public int getLastOccurance(int[] nums,int target) {
        int left = 0;
        int right = nums.length - 1;
        while(left < right -1) {
            int mid = left + (right-left)/2;
           if( nums[mid]>target) {
                right = mid;
            } else {
                left = mid;
            }
        }
        if(nums[right] == target) {
            return right;
        }
        if(nums[left] == target) {
            return left;
        }
        return -1;
    }
}

File: BinarySearch/34.Search for a Range/Solution.py
class Solution:
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """

        res = [-1,-1]
        if len(nums) == 0:
            return [-1, -1]
        if len(nums) == 1 and nums[0] == target:
            return [0,0]


        res[0] = self.__get_starting_position(nums, target)
        res[1] = self.__get_ending_position(nums, target)
        return res


    def __get_starting_position(self, nums, target):
        left = 0
        right = len(nums) - 1

        while left < right - 1:
            mid = left + (right - left) // 2
            if nums[mid] < target:
                left = mid
            else:
                right = mid
        if nums[left] == target:
            return left
        elif nums[right] == target:
            return right
        else:
            return -1
    def __get_ending_position(self, nums, target):
        left = 0
        right = len(nums) - 1

        while left < right - 1:
            mid = left + (right - left) // 2
            if nums[mid] > target:
                right = mid
            else:
                left = mid
        if nums[right] == target:
            return right
        elif nums[left] == target:
            return left
        else:
            return -1
